homeassistant/components/downloader/__init__.py

```diff
@@ -57,122 +86,143 @@ def setup(hass: HomeAssistant, config: ConfigType) -> bool:
-        """Start thread to download file specified in the URL."""
-
-        def do_download() -> None:
-            """Download the file."""
-            try:
-                url = service.data[ATTR_URL]
-
-                subdir = service.data.get(ATTR_SUBDIR)
+    lock = asyncio.Lock()
 
-                filename = service.data.get(ATTR_FILENAME)
+    async def _async_run_download(call: ServiceCall) -> None:
+        async with lock:
+            await _async_download_file(hass, call, download_path)
 
-                overwrite = service.data.get(ATTR_OVERWRITE)
-
-                if subdir:
-                    # Check the path
-                    raise_if_invalid_path(subdir)
-
-                final_path = None
-
-                req = requests.get(url, stream=True, timeout=10)
-
-                if req.status_code != HTTPStatus.OK:
-                    _LOGGER.warning(
-                        "Downloading '%s' failed, status_code=%d", url, req.status_code
-                    )
-                    hass.bus.fire(
-                        f"{DOMAIN}_{DOWNLOAD_FAILED_EVENT}",
-                        {"url": url, "filename": filename},
-                    )
+    async_register_admin_service(
+        hass,
+        DOMAIN,
+        SERVICE_DOWNLOAD_FILE,
+        _async_run_download,
+        schema=vol.Schema(
+            {
+                vol.Optional(ATTR_FILENAME): cv.string,
+                vol.Optional(ATTR_SUBDIR): cv.string,
+                vol.Required(ATTR_URL): cv.url,
+                vol.Optional(ATTR_OVERWRITE, default=False): cv.boolean,
+            }
+        ),
+    )
 
-                else:
-                    if filename is None and "content-disposition" in req.headers:
-                        match = re.findall(
-                            r"filename=(\S+)", req.headers["content-disposition"]
-                        )
+    return True
 
-                        if match:
-                            filename = match[0].strip("'\" ")
 
-                    if not filename:
-                        filename = os.path.basename(url).strip()
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload downloader."""
+    hass.services.async_remove(domain=DOMAIN, service=SERVICE_DOWNLOAD_FILE)
+    return True
 
-                    if not filename:
-                        filename = "ha_download"
 
-                    # Check the filename
-                    raise_if_invalid_filename(filename)
+async def _async_download_file(
+    hass: HomeAssistant, service: ServiceCall, download_path: str
+) -> None:
+    """Start thread to download file specified in the URL."""
+    threading.Thread(
+        target=_async_do_download, args=(hass, service, download_path)
+    ).start()
+
+
+def _async_do_download(hass: HomeAssistant, service: ServiceCall, download_path: str):
+    """Download the file."""
+    try:
+        url = service.data.get(ATTR_URL)
```

Comment:
We know what we get from the schema, let's type annotate accordingly
```suggestion
        url: str = service.data.get(ATTR_URL)
```